core/src/main/java/com/linecorp/armeria/internal/AnnotatedHttpServiceParamUtil.java

```diff
@@ -208,6 +217,135 @@ public static Object convertParameter(@Nullable String value,
+    }
+
+    /**
+     * Returns the value of the {@link Header} annotation which is specified on the {@code element} if
+     * the value is not blank. If the value is blank, it returns the name of the specified
+     * {@code nameRetrievalTarget} object which is an instance of {@link Parameter} or {@link Field}.
+     *
+     * <p>Note that the name of the specified {@code nameRetrievalTarget} will be converted as
+     * {@link CaseFormat#LOWER_HYPHEN} that the string elements are separated with one hyphen({@code -})
+     * character. The value of the {@link Header} annotation will not be converted because it is clearly
+     * specified by a user.
+     */
+    @Nullable
+    public static String findHeaderName(AnnotatedElement element, Object nameRetrievalTarget) {
+        requireNonNull(element, "element");
+        requireNonNull(nameRetrievalTarget, "nameRetrievalTarget");
+        final Header header = element.getAnnotation(Header.class);
+        if (header == null) {
+            return null;
+        }
+
+        final String value = header.value();
+        if (DefaultValues.isSpecified(value)) {
+            checkArgument(!value.trim().isEmpty(), "value: %s (expected: not blank)");
+            return value;
+        }
+        return toHeaderName(getName(nameRetrievalTarget));
+    }
+
+    /**
+     * Returns whether the specified {@code element} is annotated with a {@link Param} or {@link Header}.
+     */
+    public static boolean hasParamOrHeader(AnnotatedElement element) {
+        return element.isAnnotationPresent(Param.class) ||
+               element.isAnnotationPresent(Header.class);
+    }
+
+    /**
+     * Returns {@code true} if the specified {@code element} is not annotated with a {@link Param} nor
+     * {@link Header}.
+     */
+    public static boolean hasNeitherParamNorHeader(AnnotatedElement element) {
+        return !element.isAnnotationPresent(Param.class) &&
+               !element.isAnnotationPresent(Header.class);
+    }
+
+    private static String getName(Object element) {
+        if (element instanceof Parameter) {
+            final Parameter parameter = (Parameter) element;
+            checkArgument(parameter.isNamePresent(),
+                          "cannot obtain the name of the parameter or field automatically. " +
+                          "Please check whether your compile option includes '-parameters'. " +
+                          "If not, you need to specify parameter and header names with @%s and @%s.",
+                          Param.class.getName(), Header.class.getName());
+            return parameter.getName();
+        }
+        if (element instanceof Field) {
+            return ((Field) element).getName();
+        }
+        throw new IllegalArgumentException("cannot find the name: " + element.getClass().getName());
+    }
+
+    @VisibleForTesting
+    static String toHeaderName(String name) {
+        checkArgument(!Strings.isNullOrEmpty(name), "name");
```

Comment:
Did you mean `requireNonNull`? otherwise we need to change the second parameter which is more specific string.